### DISCUSSION TITLE
How to make similar pixels transparent while accounting transparency

### DISCUSSION THREAD

**Discussion by scarf005** [ImageMagick/ImageMagick6#238] [General] [answered]
### ImageMagick version

6.9.11

### Operating system, version and so on

ubuntu 23.04

### Description

| a | b | expected |
|--------|--------|----|
| ![a][a] | ![b][b] | ![ex][ex] |

i want to remove background (`a`) from an image (`b`). the backgrounds are identical. in other words, `a + expected = b`. now from `a.png` and `b.png`, i want to subtract `a.png` from `b.png` to get `expected` (with transparency), however [changemask command](https://imagemagick.org/Usage/compose/#changemask) 
```sh
composite a.png b.png -compose ChangeMask  res.png
```
gives me this instead because the red spray has transparency:
![res][res]

would there be any way to achieve this?

[a]: https://user-images.githubusercontent.com/54838975/236717777-242ae9f5-a9a3-49a0-86e8-3e3857b9c99e.png
[b]: https://user-images.githubusercontent.com/54838975/236717781-b403d0eb-d337-43e0-9f49-a84d2ecf922a.png
[ex]: https://user-images.githubusercontent.com/54838975/236717782-d4ddcb17-9300-42c9-9a19-6cb6348869c5.png
[res]: https://user-images.githubusercontent.com/54838975/236717941-e0a1f095-e7db-4ccf-b7ac-fc34b9b4e350.png

**Comment by snibgo** [comment, marked as answer]
I name your images `transA.png`, `transB.png` and `transExp.png`.

`transA.png` has no ICC colour profile, but the other two do have a profile. Why is that?

`transExp.png` is entirely red, with varying transparency.

>  in other words, a + expected = b. now from a.png and b.png, i want to subtract a.png from b.png to get expected (with transparency), ...

Some experimentation shows that you are wrong. Adding `transA.png` and `transExp.png` does not give transB.png.
```
magick transA.png transExp.png -compose Plus -composite x.png
```
`x.png` is not the same as `transB.png. Therefore subtracting transA from transB will not give the expected result.

However, using `over` does work:
```
magick transA.png transExp.png -compose Over -composite +depth x.png
```
`x.png` is virtually the same as `transB.png`. (If we use `depth 8`, it is exactly the same.)

So we need the inverse of `"-compose Over -composite`". For that, see [Inverse composites](http://im.snibgo.com/invcomp.htm) and [Watermarks](http://im.snibgo.com/watermark.htm).

Window syntax:
```
magick ^
  -define compose:clamp=off ^
  ( transA.png transB.png ^
    -strip -colorspace sRGB ^
    -compose MinusDst -composite ^
  ) ^
  ( transA.png ^
    -strip -colorspace sRGB ^
    ( +clone -fill Red -colorize 100 +write mpr:RED ) ^
    -compose MinusDst -composite ^
  ) ^
  -compose DivideSrc -composite ^
  +depth ^
  +write a.png ^
  -grayscale Average ^
  mpr:RED ^
  +swap ^
  -alpha off -compose CopyOpacity -composite ^
  a2.png
```
`a.png` has three channels, the calculated alpha for the R, G and B channels. They are slightly different because the input images are only 8 bits per channel. We use the average for the required single alpha value.

We use that alpha with red pixels for `a2.png`, the calculated "expected" image. How close is it to `transExp.png`?
```
magick compare -metric RMSE a2.png transExp.png NULL:

8.03125 (0.000122549)
```
This is very close.

**Reply by scarf005** [reply]
thank you for the detailed explanation! however, i should have stated it more clearly, that this was a simplified question, and that the actual image in question is not necessarily red. would the solution work for other type of overlayed images that are mixed in color (i.e transparent house)?

about ICC profile, it's because i used the first image as-is but edited on graphic tool for other images.

**Comment by snibgo** [comment]
>  ... the actual image in question is not necessarily red.

See the pages I linked to. For each pixel, if we know the colour then we can calculate the opacity. Or, if we know the opacity, we can calculate the colour. If we know neither the colour nor the opacity, an almost infinite number of solutions are possible.
